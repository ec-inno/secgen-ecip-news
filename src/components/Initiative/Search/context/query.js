import React from 'react';

import { defaultLangKey } from '../../../../../languages';

const queryInit = {
  filters: {},
  status: 'ALL',
  language: defaultLangKey,
  pagination: '0/10',
};

const queryReducer = (state, action) => {
  switch (action.type) {
    case 'paginate': {
      return {
        ...state,
        pagination: action.pagination,
      };
    }

    case 'changeStatus': {
      return {
        ...state,
        pagination: queryInit.pagination,
        status: action.status,
      };
    }

    case 'changeLanguage': {
      return {
        ...state,
        pagination: queryInit.pagination,
        language: action.language,
      };
    }

    case 'changeFilter': {
      return {
        ...state,
        pagination: queryInit.pagination,
        filters: {
          ...state.filters,
          [action.filter]: action.filterValue,
        },
      };
    }

    case 'setFilters': {
      return {
        ...state,
        pagination: queryInit.pagination,
        filters: {
          ...state.filters,
          ...action.filters,
        },
      };
    }

    case 'removeFilter': {
      const { filters } = state;
      delete filters[action.filter];

      return {
        ...state,
        pagination: queryInit.pagination,
        filters: {
          ...filters,
        },
      };
    }

    case 'reset':
    default: {
      return queryInit;
    }
  }
};

const Context = React.createContext();

export { queryInit, queryReducer, Context as default };
