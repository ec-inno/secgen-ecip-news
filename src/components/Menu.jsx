import React, { useContext } from 'react';
import classnames from 'classnames';
import { useStaticQuery, graphql, Link } from 'gatsby';

import I18nContext from '../context/I18n';

const Menu = () => {
  const { location, locale } = useContext(I18nContext);

  // Menu information about `fetched_alias` contains site folder we don't need.
  const drupalSiteFolder = process.env.GATSBY_DRUPAL_SITE_FOLDER
    ? `/${process.env.GATSBY_DRUPAL_SITE_FOLDER}/`
    : '';

  const locationParts =
    location && location.pathname
      ? location.pathname.split('/').filter(p => p)
      : [];
  const urlPath = locationParts[1];

  const data = useStaticQuery(graphql`
    query getMainMenu {
      allMenu(filter: { menu_name: { eq: "main" }, enabled: { eq: true } }) {
        edges {
          node {
            id
            title
            external
            href: fetched_alias
            enabled
            parent {
              id
            }
          }
        }
      }
    }
  `);

  const links = data.allMenu.edges.length
    ? data.allMenu.edges
        // Take essential information.
        .map(({ node }) => node)
        // Keep only first-level menu items.
        .filter(link => !link.parent)
        // Keep items in the given language because static query does not support variables at the moment.
        .filter(link => link.id.includes(`/${locale}/`))
        .map(link => ({
          id: link.id,
          label: link.title,
          href: '/' + link.href.replace(drupalSiteFolder, ''),
        }))
    : [];

  return (
    <>
      {links && links.length ? (
        <nav className="eci-menu">
          <div className="ecl-container">
            <ul className="eci-menu__list">
              {links.map(link => {
                const { id, label, href } = link;

                let classActive = '';

                if (
                  // Home page.
                  (urlPath === undefined &&
                    // Without trailing slash.
                    location.pathname.replace(/\/$/, '') === href) ||
                  // Internal pages.
                  (urlPath && href.includes(urlPath))
                ) {
                  classActive = 'eci-menu__option--is-selected';
                }

                const classNames = classnames(classActive, 'eci-menu__option');

                return (
                  <li className={classNames} key={id}>
                    <Link to={href} className="eci-menu__link ecl-link">
                      {label}
                    </Link>
                  </li>
                );
              })}
            </ul>
          </div>
        </nav>
      ) : (
        ''
      )}
    </>
  );
};

export default Menu;
