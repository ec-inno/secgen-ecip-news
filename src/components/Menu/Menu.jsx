import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Link } from 'gatsby';

const Menu = ({ items, location }) => {
  if (!items || items.length === 0) {
    return '';
  }

  const locationParts =
    location && location.pathname
      ? location.pathname.split('/').filter(p => p)
      : [];
  const urlPath = locationParts[1];

  return (
    <nav className="eci-menu">
      <div className="ecl-container">
        <ul className="eci-menu__list">
          {items.map(item => {
            const { id, label, href } = item;

            let classActive = '';

            if (
              // Home page.
              (urlPath === undefined &&
                // Without trailing slash.
                location.pathname.replace(/\/$/, '') === href) ||
              // Internal pages.
              (urlPath && href.includes(urlPath))
            ) {
              classActive = 'eci-menu__option--is-selected';
            }

            const classNames = classnames(classActive, 'eci-menu__option');

            return (
              <li className={classNames} key={id}>
                <Link to={href} className="eci-menu__link ecl-link">
                  {label}
                </Link>
              </li>
            );
          })}
        </ul>
      </div>
    </nav>
  );
};

Menu.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      label: PropTypes.string,
      href: PropTypes.string,
    })
  ),
  location: PropTypes.shape({
    pathname: PropTypes.string,
  }),
};

Menu.defaultProps = {
  items: [],
  location: {},
};

export default Menu;
