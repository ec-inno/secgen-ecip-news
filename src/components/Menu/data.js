const data = {
  bg: [
    { label: 'Начало', href: '/' },
    { label: 'Новини', href: '/news' },
    { label: 'Как работи', href: '/about' },
  ],
  en: [
    { label: 'Home', href: '/' },
    { label: 'News', href: '/news' },
    { label: 'How it works', href: '/about' },
  ],
  fr: [
    { label: 'Home (FR)', href: '/' },
    { label: 'News (FR)', href: '/news' },
    { label: 'How it works (FR)', href: '/about' },
  ],
};

export default data;
