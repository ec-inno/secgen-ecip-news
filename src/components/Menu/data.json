[
  {
    "type": "menu",
    "id": "33406188-bc68-47bd-857a-43548725498d",
    "attributes": {
      "drupal_internal__id": 2,
      "langcode": "en",
      "bundle": "menu_link_content",
      "enabled": true,
      "title": "Home",
      "description": null,
      "menu_name": "main",
      "link": {
        "uri": "internal:/",
        "title": "",
        "options": []
      },
      "external": false,
      "rediscover": true,
      "weight": 0,
      "expanded": false,
      "parent": null,
      "changed": "2019-08-23T11:22:04+00:00",
      "default_langcode": true,
      "revision_translation_affected": true,
      "fetched_alias": "/multisite/ecip/en"
    },
    "links": {
      "self": {
        "href": "https://webgate.acceptance.ec.europa.eu/multisite/ecip/en/api/menu/33406188-bc68-47bd-857a-43548725498d"
      }
    }
  },
  {
    "type": "menu",
    "id": "c317d36c-8491-405f-9b3d-1fc10046b7b3",
    "attributes": {
      "drupal_internal__id": 3,
      "langcode": "en",
      "bundle": "menu_link_content",
      "enabled": true,
      "title": "How it works",
      "description": null,
      "menu_name": "main",
      "link": {
        "uri": "entity:node/9",
        "title": "",
        "options": []
      },
      "external": false,
      "rediscover": false,
      "weight": 0,
      "expanded": false,
      "parent": null,
      "changed": "2019-08-23T11:26:10+00:00",
      "default_langcode": true,
      "revision_translation_affected": true,
      "fetched_alias": "/multisite/ecip/en/how-it-works"
    },
    "links": {
      "self": {
        "href": "https://webgate.acceptance.ec.europa.eu/multisite/ecip/en/api/menu/c317d36c-8491-405f-9b3d-1fc10046b7b3"
      }
    }
  },
  {
    "type": "menu",
    "id": "5d54a86a-9965-4649-9e00-8339fde81b20",
    "attributes": {
      "drupal_internal__id": 4,
      "langcode": "en",
      "bundle": "menu_link_content",
      "enabled": true,
      "title": "Regulatory framework",
      "description": null,
      "menu_name": "main",
      "link": {
        "uri": "entity:node/10",
        "title": "",
        "options": []
      },
      "external": false,
      "rediscover": false,
      "weight": 0,
      "expanded": false,
      "parent": "menu_link_content:c317d36c-8491-405f-9b3d-1fc10046b7b3",
      "changed": "2019-08-23T11:26:57+00:00",
      "default_langcode": true,
      "revision_translation_affected": true,
      "fetched_alias": "/multisite/ecip/en/how-it-works/regulatory-framework"
    },
    "links": {
      "self": {
        "href": "https://webgate.acceptance.ec.europa.eu/multisite/ecip/en/api/menu/5d54a86a-9965-4649-9e00-8339fde81b20"
      }
    }
  }
]
