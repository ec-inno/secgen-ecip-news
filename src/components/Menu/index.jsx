import React from 'react';
import classnames from 'classnames';
import { Link } from 'gatsby';

const Menu = ({ location }) => {
  const data = require('./data.json');

  const remove = '/multisite/ecip/';
  const locationParts =
    location && location.pathname
      ? location.pathname.split('/').filter(p => p)
      : [];
  const urlPath = locationParts[1];

  const links = data.length
    ? data
        // Keep only items which are enabled are meant to be displayed.
        .filter(link => link.attributes.enabled)
        // Keep only first-level menu items.
        .filter(link => !link.attributes.parent)
        .map(link => ({
          id: link.id,
          label: link.attributes.title,
          href: '/' + link.attributes.fetched_alias.replace(remove, ''),
        }))
    : [];

  return (
    <>
      {links && links.length ? (
        <nav className="eci-menu">
          <div className="ecl-container">
            <ul className="eci-menu__list">
              {links.map(link => {
                const { id, label, href } = link;

                let classActive = '';

                if (
                  // Home page.
                  (urlPath === undefined && location.pathname === href) ||
                  // Internal pages.
                  (urlPath && href.includes(urlPath))
                ) {
                  classActive = 'eci-menu__option--is-selected';
                }

                const classNames = classnames(classActive, 'eci-menu__option');

                return (
                  <li className={classNames} key={id}>
                    <Link to={href} className="eci-menu__link ecl-link">
                      {label}
                    </Link>
                  </li>
                );
              })}
            </ul>
          </div>
        </nav>
      ) : (
        ''
      )}
    </>
  );
};

export default Menu;
