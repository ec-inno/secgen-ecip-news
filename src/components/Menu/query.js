import { useStaticQuery, graphql } from 'gatsby';

export const useQuery = () => {
  const data = useStaticQuery(graphql`
    query getMainMenu {
      allMenu(
        filter: {
          menu_name: { eq: "main" }
          enabled: { eq: true }
          menu_has_parent: { eq: false }
        }
      ) {
        edges {
          node {
            menu_name
            id
            title
            external
            href: fetched_alias
            enabled
            menu_has_parent
          }
        }
      }
    }
  `);

  return data;
};
