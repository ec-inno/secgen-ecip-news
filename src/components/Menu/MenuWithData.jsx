import React from 'react';

import Menu from './Menu';
import { useQuery } from './query';
import { useI18nContext } from '../../context/I18n';

const MenuWithData = () => {
  const { location, locale } = useI18nContext();
  const data = useQuery();

  // Menu information about `fetched_alias` contains site folder we don't need.
  const drupalSiteFolder = process.env.GATSBY_DRUPAL_SITE_FOLDER
    ? `/${process.env.GATSBY_DRUPAL_SITE_FOLDER}/`
    : '';

  const items = data.allMenu.edges.length
    ? data.allMenu.edges
        // Take essential information.
        .map(({ node }) => node)
        // Keep items in the given language because static query does not support variables at the moment.
        .filter(link => link.id.includes(`/${locale}/`))
        .map(link => ({
          id: link.id,
          label: link.title,
          href: '/' + link.href.replace(drupalSiteFolder, ''),
        }))
    : [];

  return <Menu items={items} location={location} />;
};

export default MenuWithData;
