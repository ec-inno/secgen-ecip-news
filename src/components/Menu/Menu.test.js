import React from 'react';
import renderer from 'react-test-renderer';

import * as I18nContext from '@eci/context/I18n';
import * as query from './query';
import Menu from './MenuWithData';

describe('Menu', () => {
  it('renders correctly', () => {
    jest.spyOn(I18nContext, 'useI18nContext').mockImplementation(() => ({
      locale: 'en',
      location: { pathname: '/en' },
    }));

    jest.spyOn(query, 'useQuery').mockImplementation(() => ({
      allMenu: {
        edges: [
          {
            node: {
              menu_name: 'main',
              id: 'api/en/20a954b4-fbcf-5522-bc32-7182b255122d',
              title: 'Home',
              external: false,
              href: '#',
              enabled: true,
              menu_has_parent: false,
            },
          },
        ],
      },
    }));

    const tree = renderer.create(<Menu />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
