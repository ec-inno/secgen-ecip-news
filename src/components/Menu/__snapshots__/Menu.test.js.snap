// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly 1`] = `
<nav
  className="eci-menu"
>
  <div
    className="ecl-container"
  >
    <ul
      className="eci-menu__list"
    >
      <li
        className="eci-menu__option"
      >
        <a
          className="eci-menu__link ecl-link"
          href="/#"
        >
          Home
        </a>
      </li>
    </ul>
  </div>
</nav>
`;
